### PR TITLE
update cudf to 22.04, cuda to 11.5

### LINKED ISSUE #89: [FEA] Update Morpheus to Use Latest cuDF
**Is your feature request related to a problem? Please describe.**
As of the GA release, Morpheus is stuck on cuDF version 21.10 because several modification to the cuDF source code were necessary and these modifications cannot be applied to later releases. The modified files are contained in a patch file located [here](https://github.com/NVIDIA/Morpheus/blob/branch-22.06/cmake/deps/patches/cudf.patch) and can be summarized by the following:
1. Addition of `${CUDF_SOURCE_DIR}/include/nvtext` to the installed headers to allow Morpheus to use the C++ `SubwordTokenizer`
2. Public export of the cython `Column` class (this could possibly be removed)
3. Public export of the cython `Table` class

All of these modifications revolve around exposing public aspects of the cuDF library to external projects and were originally intended to be upstreamed after the 0.2 EA release. However, in the 21.12 release of cuDF, the `Table` class (item `#3` above) was removed completely, preventing applying the patch to any release after 21.10.

With the removal of the `Table` cython class, Morpheus will need to refactor the `morpheus.TableInfo` class and the `cudf_helpers.pyx` code.

**How the `Table` class is used in Morpheus**
Morpheus stores a message's `DataFrame` object in C++ using a `pybind11::object` wrapped in the class `IDataTable` which is then referenced by a `TableInfo` struct. To operate on the data in a `DataFrame` using C++, we need to convert the python object referenced in a `TableInfo` into a `cudf:table_view`. The `Table` cython class facilitates this conversion in C++ even though it is not used directly (we cast a pybind11 object into cython via `(PyTable*)data_frame.ptr()` and is the main reason that the `Table` class needed to be exported in patch item `#3` above).

- To convert from a `pybind11::object` -> `cudf:table_view`, we use the cudf cython method `cudf._lib.table.table_view_from_table()`
- To convert from a `cudf::table_view` -> `pybind11::object`, we use the cudf cython method `cudf._lib.utils.data_from_table_view()`

**Describe the solution you'd like**
Ultimately, the solution to this problem should allow Morpheus to utilize the most recent version of cuDF without needing a patch file or custom cuDF build while maintaining the C++ performance currently available in GA. The best solution will likely require working with the cuDF team to find alternatives and upstream any changes they see fit. Below is a rough outline of some of the expected changes:

1. Removal of a cuDF patch file and custom cuDF conda build from the Docker components
2. Update to the `IDataTable` class to store C++ objects instead of the python `DataFrame`
3. Update to the `TableInfo` class to handle slicing into the `IDataTable` using C++ functions instead of python
4. Update `cudf_helpers` to work with the latest cuDF classes and utility functions

**Describe alternatives you've considered**
While updating the `IDataTable` and `TableInfo` classes to store data in C++ instead of python would be preferred, it may not be necessary. Storing the data in C++ could significantly improve performance but may complicate the implementation. If this proves to be too difficult, alternative options that find workarounds to the removed classes while preserving the functionality would also satisfy the main goal of using the latest cuDF without modifications.

### PR DESCRIPTION
**Pull request by cwharris**
Contributes to #144 (22.06 build args).
Closes #89 (update cudf).

This PR updates cudf to 22.04 and cuda to 11.5, which is required to compile against cudf 22.04 due to decimal128 support.

Some documentation may still need updating.

### CODE REVIEW

**Comment by cwharris** [comment]
One last test failure. Runs fine locally, but complains about bash versions older than 4.4. Not sure why we wouldn't be getting this error on other PRs as well.
```
[2022-06-09T14:45:11.610Z] =================================== FAILURES ===================================
[2022-06-09T14:45:11.610Z] __________________________ TestCLI.test_autocomplete ___________________________
[2022-06-09T14:45:11.610Z] 
[2022-06-09T14:45:11.610Z] self = <test_cli.TestCLI object at 0x7fd810652430>
[2022-06-09T14:45:11.610Z] tmp_path = PosixPath('/tmp/pytest-of-unknown/pytest-0/test_autocomplete0')
[2022-06-09T14:45:11.610Z] 
[2022-06-09T14:45:11.610Z]     def test_autocomplete(self, tmp_path):
[2022-06-09T14:45:11.610Z]         runner = CliRunner()
[2022-06-09T14:45:11.610Z]         result = runner.invoke(cli.cli, ['tools', 'autocomplete', 'show'], env={'HOME': str(tmp_path)})
[2022-06-09T14:45:11.610Z] >       assert result.exit_code == 0, result.output
[2022-06-09T14:45:11.610Z] E       AssertionError: 
[2022-06-09T14:45:11.610Z] E       assert 1 == 0
[2022-06-09T14:45:11.610Z] E        +  where 1 = <Result CalledProcessError(1, ['_MORPHEUS_COMPLETE=bash_source morpheus'])>.exit_code
[2022-06-09T14:45:11.610Z] 
[2022-06-09T14:45:11.610Z] test_cli.py:149: AssertionError
[2022-06-09T14:45:11.610Z] ----------------------------- Captured stderr call -----------------------------
[2022-06-09T14:45:11.610Z] Traceback (most recent call last):
[2022-06-09T14:45:11.610Z]   File "/opt/conda/envs/morpheus/bin/morpheus", line 8, in <module>
[2022-06-09T14:45:11.610Z]     sys.exit(run_cli())
[2022-06-09T14:45:11.610Z]   File "/opt/conda/envs/morpheus/lib/python3.8/site-packages/morpheus/cli.py", line 1409, in run_cli
[2022-06-09T14:45:11.610Z]     cli(obj={}, auto_envvar_prefix='MORPHEUS', show_default=True, prog_name="morpheus")
[2022-06-09T14:45:11.610Z]   File "/opt/conda/envs/morpheus/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
[2022-06-09T14:45:11.610Z]     return self.main(*args, **kwargs)
[2022-06-09T14:45:11.610Z]   File "/opt/conda/envs/morpheus/lib/python3.8/site-packages/click/core.py", line 1050, in main
[2022-06-09T14:45:11.610Z]     self._main_shell_completion(extra, prog_name, complete_var)
[2022-06-09T14:45:11.610Z]   File "/opt/conda/envs/morpheus/lib/python3.8/site-packages/click/core.py", line 1125, in _main_shell_completion
[2022-06-09T14:45:11.611Z]     rv = shell_complete(self, ctx_args, prog_name, complete_var, instruction)
[2022-06-09T14:45:11.611Z]   File "/opt/conda/envs/morpheus/lib/python3.8/site-packages/click/shell_completion.py", line 45, in shell_complete
[2022-06-09T14:45:11.611Z]     echo(comp.source())
[2022-06-09T14:45:11.611Z]   File "/opt/conda/envs/morpheus/lib/python3.8/site-packages/click/shell_completion.py", line 326, in source
[2022-06-09T14:45:11.611Z]     self._check_version()
[2022-06-09T14:45:11.611Z]   File "/opt/conda/envs/morpheus/lib/python3.8/site-packages/click/shell_completion.py", line 314, in _check_version
[2022-06-09T14:45:11.611Z]     raise RuntimeError(
[2022-06-09T14:45:11.611Z] RuntimeError: Shell completion is not supported for Bash versions older than 4.4.
```

**Comment by cwharris** [comment]
I assume the test failure has to do with the fact that we are updating the build images to centos, and that centos does not have a compatible shell for `click`.

**Comment by pdmack** [comment]
Somehow in the docker build it's trying to install multiple cudatoolkit
```
#19 96.91 RuntimeError: Solver could not find solution.Mamba failed to solve:
#19 96.91  - versioneer-518
#19 96.91  - libgcc-ng >=12
#19 96.91  - scikit-build >=0.12
#19 96.91  - rapidjson 1.1.*
#19 96.91  - cudatoolkit 11.4.*
#19 96.91  - librdkafka 1.6.1.*
#19 96.91  - cython >=0.29,<0.30
#19 96.91  - pip
#19 96.91  - neo 22.04.*
#19 96.91  - __glibc >=2.17
#19 96.91  - cudatoolkit >=11.5,<12
#19 96.91  - cudf 22.04.*
#19 96.91  - python 3.8.*
#19 96.91  - libstdcxx-ng >=12
#19 96.91
#19 96.91 with channels:
#19 96.91  - rapidsai
#19 96.91  - nvidia
#19 96.91  - nvidia/label/dev
#19 96.91  - conda-forge
#19 96.91
#19 96.91 The reported errors are:
#19 96.91 - Encountered problems while solving:
#19 96.91 -   - cannot install both cudatoolkit-11.4.2-h00f7ccd_9 and cudatoolkit-11.7.0-hd8887f6_10
```

File: ci/Jenkinsfile [hunk 1]
```diff
@@ -19,7 +19,7 @@ pipeline {
                     }
                     agent {
                         docker {
-                            image 'gpuci/rapidsai-driver:21.10-cuda11.4-devel-ubuntu20.04-py3.8'
+                            image 'gpuci/rapidsai-driver:22.04-cuda11.5-devel-centos7-py3.8'
```
Comment (mdemoret-nv) [inline ci/Jenkinsfile:22]: What is the reason for changing from ubuntu to centos here?

File: ci/Jenkinsfile [hunk 2]
```diff
@@ -43,7 +43,7 @@ pipeline {
                     }
                     agent {
                         docker {
-                            image 'gpuci/rapidsai:21.10-cuda11.4-devel-ubuntu20.04-py3.8'
+                            image 'gpuci/rapidsai:22.04-cuda11.5-devel-centos7-py3.8'
```
Comment (mdemoret-nv) [inline ci/Jenkinsfile:46]: Can we use env variables in the docker image? Would be better to have it as `image 'gpuci/rapidsai:${RAPIDS_VER}-cuda${CUDA_VER}-devel-centos7-py${PYTHON_VER}'`

File: docker/Dockerfile [hunk 3]
```diff
@@ -18,7 +18,7 @@
 
 # Args used in FROM commands must come first
 ARG FROM_IMAGE="gpuci/miniforge-cuda"
-ARG CUDA_VER=11.4
+ARG CUDA_VER=11.5
 ARG LINUX_DISTRO=ubuntu
 ARG LINUX_VER=20.04
```
Comment (mdemoret-nv) [inline docker/Dockerfile:23]: Do we need to update the default linux distro here to use centos?

File: docker/build_container.sh [hunk 4]
```diff
@@ -23,10 +23,10 @@ DOCKER_EXTRA_ARGS=${DOCKER_EXTRA_ARGS:-""}
 
 # Build args
 FROM_IMAGE=${FROM_IMAGE:-"gpuci/miniforge-cuda"}
-CUDA_VER=${CUDA_VER:-11.4}
+CUDA_VER=${CUDA_VER:-11.5}
 LINUX_DISTRO=${LINUX_DISTRO:-ubuntu}
```
Comment (mdemoret-nv) [inline docker/build_container.sh:27]: Again, does this need to be updated to centos?

File: docker/conda/environments/cuda11.5_dev.yml [hunk 5]
```diff
@@ -29,10 +28,10 @@ dependencies:
     - ccache>=3.7
     - cmake=3.22
     - cuda-nvml-dev=11.4
```
Comment (mdemoret-nv) [inline docker/conda/environments/cuda11.5_dev.yml:30]: No `cuda-nvml-dev=11.5`?

File: docker/conda/environments/cuda11.5_ci.yml [hunk 6]
```diff
@@ -18,6 +18,7 @@ name: morpheus
 channels:
     - conda-forge
 dependencies:
+    - bash=5.1.16
```
Comment (mdemoret-nv) [inline docker/conda/environments/cuda11.5_ci.yml:21]: Can you add a comment on why this version is required?

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
Couple of things:

- These lines: https://github.com/nv-morpheus/Morpheus/pull/148/files#diff-c42a677e35a15854d339a2f6469a2f1a21fec419e1d72c044a41ded0a14fde84R63-R75 arent needed with cudf removed.
- How come only the single `nvtext` file for the subword tokenizer was included instead of every `nvtext` file?

**Comment by cwharris** [comment]
> How come only the single nvtext file for the subword tokenizer was included instead of every nvtext file?
Morpheus only requires the one header, and cudf team has made it clear that nvtext will be exposed in 22.08. Including more files is unnecessary and will only enable us to make deviations that we'd need to fix later.

File: ci/Jenkinsfile (hunk 2 above)
Comment (cwharris) [reply ci/Jenkinsfile:46]: we don't change this often enough to matter, and there's a chance the image won't be available, so best to let people search for a good image and just copy paste.

File: docker/build_container.sh (hunk 4 above)
Comment (cwharris) [reply docker/build_container.sh:27]: Again, I don't think this is necessary.

File: docker/conda/environments/cuda11.5_dev.yml (hunk 5 above)
Comment (cwharris) [reply docker/conda/environments/cuda11.5_dev.yml:30]: iirc, cuda-nvml-dev>=11.5,<11.6 wouldn't solve, but I can try again.

Comment (cwharris) [reply docker/conda/environments/cuda11.5_dev.yml]: we discussed offline that cuda-nvml-dev wouldn't solve when providing the `nvidia` conda channel, and instead we needed to add a specific `nvidia/label/cuda-11.5.2` channel.

**Review by mdemoret-nv — APPROVED** [review]
Looks good now.

**Comment by mdemoret-nv** [comment]
@gpucibot merge

**Pull request merged**